src/tools/recvfuzz/recvfuzz.cpp

```diff
@@ -570,423 +560,447 @@ bool WriteClientPacket(
+            WriteCryptoFrame(
+                &PayloadLength,
+                BufferSize,
+                Buffer + *HeaderLength,
+                ClientContext,
+                PacketParams);
+        }
     }
 
+    PayloadLength += GetRandom<uint8_t>(64); // More random padding
+
+    *PacketLength = *HeaderLength + PayloadLength + CXPLAT_ENCRYPTION_OVERHEAD;
+    CXPLAT_FRE_ASSERT(*PacketLength + PacketNumberLength < BufferLength);
     QuicVarIntEncode2Bytes(
-        PacketNumberLength + FrameBufferLength + CXPLAT_ENCRYPTION_OVERHEAD,
+        PacketNumberLength + PayloadLength + CXPLAT_ENCRYPTION_OVERHEAD,
         Buffer + PayloadLengthOffset);
-    *HeaderLength = *PacketLength;
-
-    CxPlatCopyMemory(Buffer + *PacketLength, FrameBuffer, FrameBufferLength);
-    *PacketLength += FrameBufferLength;
-    *PacketLength += CXPLAT_ENCRYPTION_OVERHEAD;
-
-    return true;
 }
 
-void fuzzPacket(uint8_t* Packet, uint16_t PacketLength) {
-    uint8_t numIteration = (uint8_t)GetRandom(256);
-    for(int i = 0; i < numIteration; i++){
-        Packet[GetRandom(PacketLength)] = (uint8_t)GetRandom(256);
+//
+// Finalizes the packet number, encryption, and header protection.
+//
+void FinalizePacket(
+    _Out_writes_(PacketLength)
+        uint8_t* Packet,
+    _In_ uint16_t PacketLength,
+    _In_ uint16_t HeaderLength,
+    _In_ uint64_t PacketNumber,
+    _In_ PacketParams* PacketParams,
+    _In_opt_ TlsContext* ClientContext
+    )
+{
+    uint8_t* DestCid = Packet + sizeof(QUIC_LONG_HEADER_V1);
+    QUIC_PACKET_KEY_TYPE KeyType = QuicPacketTypeToKeyTypeV1((uint8_t)PacketParams->PacketType);
+
+    QUIC_PACKET_KEY* WriteKey = nullptr;
+    if (PacketParams->Mode == 0) {
+        MUST_SUCCEED(
+            QuicPacketKeyCreateInitial(
+                FALSE,
+                &HkdfLabels,
+                InitialSalt.Data,
+                PacketParams->DestCidLen,
+                (uint8_t*)DestCid,
+                nullptr,
+                &WriteKey));
+    } else {
+        if (ClientContext->State.WriteKeys[0] == nullptr) {
+            MUST_SUCCEED(
+                QuicPacketKeyCreateInitial(
+                    FALSE,
+                    &HkdfLabels,
+                    InitialSalt.Data,
+                    PacketParams->DestCidLen,
+                    (uint8_t*)DestCid,
+                    &ClientContext->State.ReadKeys[0],
+                    &ClientContext->State.WriteKeys[0]));
+            ClientContext->State.ReadKey = QUIC_PACKET_KEY_INITIAL;
+            ClientContext->State.WriteKey = QUIC_PACKET_KEY_INITIAL;
+        }
+        WriteKey = ClientContext->State.WriteKeys[KeyType];
+    }
+    uint8_t Iv[CXPLAT_IV_LENGTH];
+    QuicCryptoCombineIvAndPacketNumber(
+        WriteKey->Iv, (uint8_t*)&PacketNumber, Iv);
+
+    CxPlatEncrypt(
+        WriteKey->PacketKey,
+        Iv,
+        HeaderLength,
+        Packet,
+        PacketLength - HeaderLength,
+        Packet + HeaderLength);
+
+    uint8_t HpMask[16];
```

Comment:
This magic number (and several others) seem non-obvious to me.

Reply:
It's a crypto magic number. I don't have a good define for you. It's the same in QUIC.